multipledispatch/dispatcher.py

```diff
@@ -169,17 +174,26 @@ def add(self, signature, func, on_ambiguity=ambiguity_warn):
                                 (typ, str_sig, self.name))
 
         self.funcs[signature] = func
-        self.reorder(on_ambiguity=on_ambiguity)
         self._cache.clear()
 
+        try:
+            del self._ordering
+        except AttributeError:
+            pass
+
+    @property
+    def ordering(self):
+        ordering = getattr(self, '_ordering', None)
+        if ordering is None:
+            ordering = self.reorder()
+        return ordering
```

Comment:
I suspect that the following is marginally faster:

```python
try:
    return self._ordering
except AttributeError:
    return self.reorder()
```

Asking foregiveness tends to introduce less overhead in Python than asking permission in the common case where permission is usually granted.

This may not be relevant though.  I haven't done benchmarks in this particular case.